server/TenantPipeline/resources/lambda-prepare-deploy.py

```diff
@@ -0,0 +1,268 @@
+            return artifact
+            
+    raise Exception('Input artifact named "{0}" not found in event'.format(name))
+
+def get_template_url(s3, artifact, file_in_zip):
+    """Gets the template artifact
+    
+    Downloads the artifact from the S3 artifact store to a temporary file
+    then extracts the zip and returns the file containing the CloudFormation
+    template.
+    
+    Args:
+        artifact: The artifact to download
+        file_in_zip: The path to the file within the zip containing the template
+        
+    Returns:
+        The CloudFormation template as a string
+        
+    Raises:
+        Exception: Any exception thrown while downloading the artifact or unzipping it
+    
+    """
+    tmp_file = tempfile.NamedTemporaryFile()
+    bucket = artifact['location']['s3Location']['bucketName']
+    print(bucket)
+
+    key = artifact['location']['s3Location']['objectKey']
+    print(key)    
+    with tempfile.NamedTemporaryFile() as tmp_file:
+        s3.download_file(bucket, key, tmp_file.name)
+        with zipfile.ZipFile(tmp_file.name, 'r') as zip:
+            extracted_file = zip.extract(file_in_zip, '/tmp/')
+            s3.upload_file(extracted_file, bucket, file_in_zip)
+            template_url =''.join(['https://', bucket,'.s3.amazonaws.com/',file_in_zip])
+            return template_url  
+  
+def put_job_success(job, message):
+    """Notify CodePipeline of a successful job
+    
+    Args:
+        job: The CodePipeline job ID
+        message: A message to be logged relating to the job status
+        
+    Raises:
+        Exception: Any exception thrown by .put_job_success_result()
+    
+    """
+    print('Putting job success')
+    print(message)
+    code_pipeline.put_job_success_result(jobId=job)
+  
+def put_job_failure(job, message):
+    """Notify CodePipeline of a failed job
+    
+    Args:
+        job: The CodePipeline job ID
+        message: A message to be logged relating to the job status
+        
+    Raises:
+        Exception: Any exception thrown by .put_job_failure_result()
+    
+    """
+    print('Putting job failure')
+    print(message)
+    code_pipeline.put_job_failure_result(jobId=job, failureDetails={'message': message, 'type': 'JobFailed'})
+
+def get_user_params(job_data):
+    """Decodes the JSON user parameters and validates the required properties.
+    
+    Args:
+        job_data: The job data structure containing the UserParameters string which should be a valid JSON structure
+        
+    Returns:
+        The JSON parameters decoded as a dictionary.
+        
+    Raises:
+        Exception: The JSON can't be decoded or a property is missing.
+        
+    """
+    try:
+        # Get the user parameters which contain the stack, artifact and file settings
+        user_parameters = job_data['actionConfiguration']['configuration']['UserParameters']
+        decoded_parameters = json.loads(user_parameters)
+            
+    except Exception:
```

Comment:
recommend catching a more specific `Exception` as per best practices: https://www.learnpython.dev/03-intermediate-python/40-exceptions/70-best-practices/#don-t-catch-exception

Looks like this might be the exception that `json.loads(...)` raises: https://www.learnpython.dev/03-intermediate-python/40-exceptions/70-best-practices/#don-t-catch-exception

Reply:
Makes sense. I will have a look. We have been pretty liberal on this front so far due to lack of unit tests.